### ISSUE TITLE
Simple authentication (managed context)

### ISSUE THREAD

**Issue by brillout** [reframejs/wildcard-api#59]
The idea is to allow the server to modify the context:

~~~js
server.login = async function(username, password) {
   if (await invalidCredentials(username, password)) {
     return;
   }

   // `this.user` is persisted
   this.user = {
     username,
   };

   // The user is now logged-in!
};

server.getUserPosts = async function() {
  // `this.user` was set by a previous `server.login` call
  const {user} = this;

  if (!user) {
    // Not logged-in
    return;
  }

  const posts = await db.getUserPosts(user.id);

  // ...
};
~~~

This could be called "managed context", and it would enable simple authentication.

Full example:
~~~js
const { server, setSecretKey } = require('@wildcard-api/server');
const db = require('./path/to/db');
const hashPassword = require('./path/to/hash-password');

const MY_SCRET_KEY = 'a-secret-string';
setSecretKey(MY_SCRET_KEY);

server.login = async function({username, password}) {
  if (!username || !password)  return;

  const user = await db.getUser(username);
  if (!user || user.passwordHashed !== hashPassword(password)) {
    return {wrongCredentials: true};
  }

  // This will persist!
  this.user = {
    id: user.id
    username: user.username,
    firstName: user.firstName,
  };
};

server.getUserPosts = async function () {
  if (!this.user)  return;

  // `this.user` has been set by a previous `server.login` call!
  const userId = this.user.id;

  const posts = await db.getUserPosts(userId);
  return posts;
};

/* How does it work?

Wildcard persists the context by using two cookies:
 - One cookie that holds the `this.user` data.
   (Without `HttpOnly` so that the client-side code can read the data.)
 - A second cookie that holds a server-side generated signature with `MY_SCRET_KEY`.
   (With `HttpOnly` so that no browser-side JS attack can steal the credentials.)

*/
~~~

~~~js
// Browser-side

import { server, client } from '@wildcard-api/client';
import assert from 'assert';
import React from 'react';
import ReactDOM from 'react-dom';

render();

function render() {
  ReactDOM.render(
    View,
    document.querySelector('#view')
  );
}

async function login() {
  // Wasn't logged-in
  assert(!client.user);

  await server.login({
    username: 'brillout',
    password: 'my-dog-fifi',
  });

  // Is now logged-in
  assert(client.user);

  // Re-render
  render();
}

function logout() {
  // Removing context data is allowed on the client-side
  delete client.user;

  // Re-render
  render();
}

function hackWildcard() {
  // Altering context data is forbidden on the client-side
  let err;
  try {
    // Nope, this won't work
    client.user = {username: 'super-admin-user'};
  } catch(err_) { err = err_ }
  assert(err);
}

function View() {
  if (!client.user)  {
    return <>
      <div>
        You are logged out.
      </div>
      <button type="button" onclick={login}>Login</button>
    </>;
  } else {
    return <>
      <div>
        Welcome {client.user.firstName}.
      </div>
      <button type="button" onclick={logout}>Logout</button>
    </>;
  }
}
~~~

**Comment by brillout** [comment]
@Johnson444 We can implement this together, if that's something you'd be interested in doing. I've cleaned up the source code, it's now all TypeScript, and test coverage is close to 100% which makes developing super comfy.

**Comment by michie1** [comment]
https://github.com/reframejs/wildcard-api/pull/58#issuecomment-711412477
> I'm curious, you mentioned a while ago using `getContext` to set cookies, are you still using that approach? What do you think of #59?
I'm only using it in a small inactive pet project, but yes. Besides putting data in the context from Express, I'm also putting functions it in. It works fine. I think it's nicer to assign the those functions to the context than to the `server` object. If you read the code you would think the functions are provided by the library, because `server` is imported from the library.

I would prefer to move Wildcard sessions to its own plugin/library, because I would prefer to handle those things myself, although it's definitely a nice idea to showcase Wildcard and have some authentication running quickly.

**Comment by brillout** [comment]
The plan is to actually integrate this with an interface that could eventually become a public API. For now the interface would be private but you're right maybe it should be exposed right at the beginning.

**Comment by brillout** [comment]
>  I think it's nicer to assign the those functions to the context than to the server object. 

Interesting. So basically you prefer to define server endpoints dynamically on each API request, is that correct? Personally, I find it simpler to reason about static functions + a context object. But maybe you are seeing something I'm not seeing.

**Comment by michie1** [comment]
> > I think it's nicer to assign the those functions to the context than to the server object.
> 
> Interesting. So basically you prefer to define server endpoints dynamically on each API request, is that correct? Personally, I find it simpler to reason about static functions + a context object. But maybe you are seeing something I'm not seeing.

No, with "those" functions I don't mean the endpoint functions, but the helper functions to interact with Express, like `setCookies()` or `getHeaders()`.

**Comment by brillout** [comment]
The `getApiHttpResponse()` function could be used to control the context:

~~~js
import express from "express";
import { getApiHttpResponse } from '@wildcard-api/server';

const app = express();

app.all('/_wildcard_api/*', async (req, res) => {
  const context = {
    user: getAuthUser(req.headers),
  };

  const {url, method, body} = req;
  const {httpResponse, endpointResult, endpointName} = await getApiHttpResponse({url, method, body}, context);

  if( endpointName==='login' && endpointResult.success ){
    setCookies(req, endpointResult.user);
  }

  // `httpResponse` contains the HTTP encoded equivalent of `endpointResult`
  res.status = httpResponse.statusCode;
  res.type = httpResponse.contentType;
  res.send(httpResponse.body);
});
~~~

This gives full control over to the user.

@michie1 Would that work for you?

> I would prefer to handle those things myself

Why? Are you using an Express middleware that does this for you?

**Comment by michie1** [comment]
> @michie1 Would that work for you?
> 
> > I would prefer to handle those things myself
> 
> Why? Are you using an Express middleware that does this for you?

Different applications need different types of authentication. I think it's better if Wildcard is providing an authentication option (as a plugin) and not enforcing anything. I'm using my own Express middleware function that puts helper functions in the context object.

**Comment by brillout** [comment]
Wildcard Sessions will be optional and you'll have more control over the context/auth than what currently is the case. 

> I'm using my own Express middleware

Can you imagine yourself removing your custom Express middleware and use Wildcard Sessions instead? If not, why not? You said you prefer to handle these things yourself and I'm left to wonder why :-).

**Comment by michie1** [comment]
> Wildcard Sessions will be optional and you'll have more control over the context/auth than what currently is the case.

Cool.

> > I'm using my own Express middleware
> 
> Can you imagine yourself removing your custom Express middleware and use Wildcard Sessions instead? If not, why not? You said you prefer to handle these things yourself and I'm left to wonder why :-).

Well, I might use it for simple applications, but like I said, different applications need different type of authentication. Maybe I don't want to use a password for authentication, but I want have an login link send by email, use 2FA or SSO. It's difficult to compare wildcard sessions that is not there yet versus all the other existing options.

**Comment by brillout** [comment]
Wildcard Sessions only does session management; you can use it with any auth strategy:

~~~js
// Username + password
server.login = function(username, password) {
  if (await invalidCredentials(username, password)) return;
  this.user = { username }; // Wildcard will automatically persist `this.user` by using HTTP cookies
};

// OAuth
server.oauthCallback = function(userProfile) {
   // At the end of the OAuth flow, we save the user information to the context object
   this.user = userProfile; // Wildcard will automatically persist `this.user` by using HTTP cookies
};

// Etc.
~~~

Instead of dealing with cookies you simply deal with reading/writing the context object.

**Comment by ggoodman** [comment]
I would suggest that if you're looking to check the authenticity of the cookie, a library like https://npm.im/@hapi/iron is the gold standard for this sort of thing. I see that my colleagues at Auth0 reach for this library for similar uses-cases.

**Comment by brillout** [comment]
> I would suggest that if you're looking to check the authenticity of the cookie, a library like https://npm.im/@hapi/iron is the gold standard for this sort of thing. I see that my colleagues at Auth0 reach for this library for similar uses-cases.

Thanks for the pointer. Wildcard [is using HMAC-SHA256](https://github.com/brillout/wildcard-api/blob/b7ca05799dcfc05e4f2c1a4e1a758a72e56224c9/src/context/server/cookie-management.ts#L143-L151) which is old, battle-tested, and assailed against by researchers. So we should be good here.
